### DISCUSSION TITLE
Possibility of making protected/encrypted levels/worlds

### DISCUSSION THREAD

**Discussion by Wohlstand** [Wohlstand/TheXTech#228] [Locked]
Because some level and episode makers want to protect their creations from possible cheating, they want to have the sort of encryption.

However:
- Having SMBX-38A being tricky and closed-source, all encryption keys were stored internally inside the executable, they aren't making the real sort of protection, and also, making a lot of troubles even for episode creators (for example, YaveYu has lost sources for the bunch of his episodes, and therefore, it's tricky to decrypt them)

Instead of encryption, there would be a sort of obfuscation that could make it majorly harder to be edited/inspected/cheated but should be fine to be played normally.

My idea is:
- Adding separated binary format for level, world, and gamesave files, they can't be just inspected/edited in a text and will be fully non-understandable using hex editor except cases user know the detailed specification of the file
- Make the semi-encryption and obfuscation of all data: all human-friendly stuff being encrypted using the password that the creator will make, so, this only password would decrypt this section. Without this section, file data will be extremely hard for inspection/editing. So, having such encrypted files, disallow open them by Editor and require to type the password to decrypt the section of lists.
- Additionally, have the special archive format to store episode data inside it (especially for making tricker way to edit, use own format instead of ordinary zip/7z which will be also supported with PhysFS)

The thing in all terms won't protect from hackers, programmers, and experienced users, but will protect from the rest of ordinary users.

**EDIT:** Forgot to tell the note: I made this topic for two purposes:
- Ask the general opinion of others, having different opinions and given points for the stuff
- The kick for this was a question of one of the users who DMed me about the encryption stuff because of the necessity to prevent cheating against secrets finding.

**Comment by 0lhi** [comment]
This is a bad idea. And I am quite honestly shocked it's even up for debate.

Anyone who's familiar with Eclipsed/DawnTimeTV knows that 99% of Level Designers don't know what they are doing. Unlike with Super Mario Maker, there is no quality control on who can upload Levels. A lot of episodes out there are straight up impossible to finish without manual bugfixes. Making them artificially unfixable for a large userbase will lead to unnecessary frustration when investing time into an episode that cannot be beaten. Same for levels where the solution is not intuitive at all (Commercial Games have that problem, too, but here Walkthroughs can help. For SMBX, the Editor often serves the same purpose.)

Another issue is that implementing such a setup would take resources that could be better used elsewhere. We currently have 110 Open Issues on top of several discussions, many of which having a different idea about what TheXTech should be and where it should go. I think, in general, it would be good to have a discussion about what our priorities are, and what is better to be left with other projects (Moondust, ReInvent etc...).

One thing is certain though: The open nature of TheXTech is it's biggest appeal. Anything that aims to make things obfuscated and harder to access should be rejected outright. That kind of proprietary mindset can stick with 38A.

**Comment by LoveBodhi** [comment]
This is actually NOT a thing that should be done! It can still be hacked!

**Comment by Wohlstand** [comment]
We had discussed something more detailed at Discord, I want to write here the summary:
- This idea is generally bad as this thing just hurts a big amount of ordinary people than cheaters, mainly, because there is a lot of bad-designed episodes with mistakes done and users won't fix them by theme selves and the episode becomes garbage because it's unbeatable because of a bug
- The general idea of the project is a community sandbox where people have their creativity shared with each other and everybody can freely contribute to the work of each other
- We are not intended to produce any proprietary crap (things like encryption/obfuscation were mainly related to proprietary crap generally)

So, I marked it as "wontfix" tag, so, leave this discussion being publically available to let everyone see that we are DON'T plan to add any sort of content encryptions.
